tools/cjxl_ng_main.cc

```diff
@@ -295,61 +258,99 @@ namespace {
   }
   if (fclose(file) != 0) {
-    fprintf(stderr,
-            "Could not close %s\n"
-            "Error: %s",
-            filename, strerror(errno));
+    std::cerr << "Could not close file: " << filename << std::endl
+              << "Error: " << strerror(errno) << std::endl;
     return false;
   }
   return true;
 }
 
-bool ProcessTristateFlag(const char* flag_name, const bool flag_value,
-                         JxlEncoderFrameSettings* frame_settings,
-                         JxlEncoderFrameSettingId encoder_option) {
-  gflags::CommandLineFlagInfo flag_info =
-      gflags::GetCommandLineFlagInfoOrDie(flag_name);
-  if (!flag_info.is_default) {
-    JxlEncoderFrameSettingsSetOption(frame_settings, encoder_option,
-                                     static_cast<int32_t>(flag_value));
+void SetFlagFrameOptionOrDie(const char* flag_name, int32_t flag_value,
+                             JxlEncoderFrameSettings* frame_settings,
+                             JxlEncoderFrameSettingId encoder_option) {
+  if (JXL_ENC_SUCCESS != JxlEncoderFrameSettingsSetOption(
+                             frame_settings, encoder_option, flag_value)) {
+    std::cerr << "Setting encoder option from flag -- " << flag_name
+              << "failed." << std::endl;
+    exit(EXIT_FAILURE);
   }
-  return true;
 }
 
-jxl::Status LoadInput(const char* filename_in,
-                      jxl::extras::PackedPixelFile& ppf) {
+void SetDistanceFromFlags(JxlEncoderFrameSettings* jxl_encoder_frame_settings,
+                          const jxl::extras::Codec& codec) {
+  bool distance_set =
+      !gflags::GetCommandLineFlagInfoOrDie("distance").is_default;
+  bool quality_set = !gflags::GetCommandLineFlagInfoOrDie("quality").is_default;
+
+  if (distance_set && quality_set) {
+    std::cerr << "Must not set both --distance and --quality." << std::endl;
+    exit(EXIT_FAILURE);
+  }
+  if (distance_set) {
+    if (JXL_ENC_SUCCESS != JxlEncoderSetFrameDistance(
+                               jxl_encoder_frame_settings, FLAGS_distance)) {
+      std::cerr << "Setting --distance parameter failed." << std::endl;
+      exit(EXIT_FAILURE);
+    }
+    return;
+  }
+  if (quality_set) {
+    double distance = FLAGS_quality >= 100 ? 0.0
+                      : FLAGS_quality >= 30
+                          ? 0.1 + (100 - FLAGS_quality) * 0.09
+                          : 6.4 + pow(2.5, (30 - FLAGS_quality) / 5.0) / 6.25;
+    if (JXL_ENC_SUCCESS !=
+        JxlEncoderSetFrameDistance(jxl_encoder_frame_settings, distance)) {
+      std::cerr << "Setting --quality parameter failed." << std::endl;
+      exit(EXIT_FAILURE);
+    }
+    return;
+  }
+  // No flag set, but input is JPG or GIF: Use distance 0 default.
+  if (codec == jxl::extras::Codec::kJPG || codec == jxl::extras::Codec::kGIF) {
+    if (JXL_ENC_SUCCESS !=
+        JxlEncoderSetFrameDistance(jxl_encoder_frame_settings, 0.0)) {
+      std::cerr << "Setting '100% quality' default for GIF or JPEG input."
```

Comment:
This should probably be in the `else` branch of that test. Also I don't like `100% quality`, it's better to say lossless.

Reply:
Thanks for spotting. Fixed.